src/niftyone/figures/generator.py

```diff
@@ -92,10 +107,27 @@ def __call__(
         overwrite: bool,
     ) -> None:
         # Filters by entity (via string query)
-        table_idxes = table.ent.query(self.query).index
-        for idx in table_idxes:
-            record = table.nested.loc[idx]
-            self.generate(record=record, out_dir=out_dir, overwrite=overwrite)
+        # First query is for main image, subsequent are for overlays
+        query_dfs = [table.ent.query(q) for q in self.queries]
+        indexed_dfs = [
+            (
+                df.assign(table_index=df.index)
+                .reset_index(drop=True)
+                .loc[:, ["table_index"] + self.join_entities]
+                .rename(columns={"table_index": f"index_{ii}"})
+            )
+            for ii, df in enumerate(query_dfs)
+        ]
+        joined = reduce(
+            lambda left, right: pd.merge(
+                left, right, on=self.join_entities, how="outer"
+            ),
+            indexed_dfs,
+        )
+        indices = [joined[f"index_{ii}"].values for ii in range(len(self.queries))]
+
+        records = [table.nested.loc[ind] for inds in zip(*indices) for ind in inds]
```

Comment:
Ahh I played around with this one for a little while to get it working. Ultimately, we need the record from the original table since we make use of the filepath there. I believe this should still be similar to the original code, just with all the joins / merges.